### PR TITLE
Reproducible development environment with Nix

### PR DESCRIPTION
**Pull request by octalsrc**
This adds a [`cuanto` Nix derivation](https://github.com/cuplv/cuanto/compare/dev-environment?expand=1#diff-5712e736e0de6ba170577f8472c398e9) which specifies `cuanto`'s external system-installed dependencies and allows quick entry into a complete development environment (as explained in [CONTRIBUTING.md](https://github.com/cuplv/cuanto/compare/dev-environment?expand=1#diff-6a3371457528722a734f3c51d9238c13)):

    # Clone the cuanto project
    $ git clone https://github.com/cuplv/cuanto
    
    # Enter the cuanto build environment
    $ cd cuanto
    $ nix-shell
    
    # Test the build environment
    [nix-shell:~/cuanto]$ sbt test

The `CONTRIBUTING.md` doc also instructs you to add `cuplv-nixpkgs` to your Nix overlays.  This is not necessary for any of `cuanto`'s current dependencies, but is used by the `apron` branch currently under PR (for the `japron` package).  This overlay would be a convenient way for CUPLV researchers to quickly make their own projects available for integration into `cuanto` in the future.

## Usage: Adding new dependencies

To add a new dependency (with a package existing in either the main [`nixpkgs`](https://github.com/nixos/nixpkgs) or [`cuplv-nixpkgs`](https://github.com/cuplv/cuplv-nixpkgs), edit `default.nix`.  Add the package's name to the *derivation inputs* (the comma-separated list in curly braces at the top of the file) and to the *build inputs* (the space-separated list in square brackets assigned to `buildInputs`).

For an example of adding a dependency, see [this commit](https://github.com/cuplv/cuanto/commit/47c3e1e4cd3ce0e863df5b914d6cd7a7a11f324c) for setting up `japron`.

## Travis builds

Additionally, this PR modifies the Travis configuration so that our CI builds use the same environment.  This way, adding new dependencies to your development environment by editing `default.nix` adds them to the Travis build environment as well.

## System class-path for SBT

Finally, this PR configures SBT to search for unmanaged jars in `$CLASSPATH`, the standard environment variable which points to all system-installed java libraries (in particular this variable is set by `nix-shell` to point to all jars from the specified dependencies).  It is necessary to use system-installed jars when they depend on native libraries, as is the case for `japron`.

### CODE REVIEW

**Comment by jpw234** [comment]
https://github.com/NixOS/nix/issues/1203

File: build.sbt [hunk 1]
```diff
@@ -71,6 +72,12 @@ lazy val root = (project in file(".")).
     resolvers += sonatypeResolver,
     libraryDependencies += scalaSMTLIB,
 
+    // Tell sbt to look at $CLASSPATH for unmanaged jars.  This is
+    // where system-installed jars (such as jars that depend on native
+    // libraries) are usually found.
+    unmanagedClasspath in Compile ++= systemClasspath,
```
Comment (bechang) [inline build.sbt]: We decided to use something like `cuantoClasspath` or `nixClasspath`.

**Comment by bechang** [comment]
@jpw234 @ftc Can one of you guys at to `CONTRIBUTING.md`.

**Comment by octalsrc** [comment]
I have changed the SBT configuration to search in `$CUANTO_CLASSPATH` for system-installed jars, and provided a utility in `default.nix` that sets this variable correctly for dependencies.

I've tested this change for `japron` on the [`japron-test` branch](https://github.com/cuplv/cuanto/commit/51cefe7292ff025e2892bb4d915c42c8ecc5e6b7)

**Comment by octalsrc** [comment]
I have changed the SBT configuration to search in `$CUANTO_CLASSPATH` for system-installed jars, and provided a utility in `default.nix` that sets this variable correctly for dependencies.

I've tested this change for `japron` on the [`japron-test` branch](https://github.com/cuplv/cuanto/commit/8208068595605d0e2a9935177966182d0bd36196)

**Review by jpw234 — APPROVED** [review]
Looks good and works on my (abnormal) system

**Review by tianhan0 — COMMENTED** [review]
Confirmed that `japron-test` passed `sbt test` on Mac

**Comment by octalsrc** [comment]
As discussed, the Windows troubleshooting documentation will be its own PR, addressing Issue #57.

**Pull request merged**